sector-builder/src/helpers/add_piece.rs

```diff
@@ -39,41 +38,82 @@ pub fn add_piece(
 
     let dest_sector_id = opt_dest_sector_id
         .ok_or(())
-        .or_else(|_| provision_new_staged_sector(sector_mgr, &mut sector_builder_state))?;
-
-    if let Some(s) = sector_builder_state.staged.sectors.get_mut(&dest_sector_id) {
-        let piece_lengths: Vec<_> = s.pieces.iter().map(|p| p.num_bytes).collect();
-
-        let (expected_num_bytes_written, mut chain) =
-            get_aligned_source(piece_file, &piece_lengths, piece_bytes_len);
-
-        sector_store
-            .manager()
-            .write_and_preprocess(&s.sector_access, &mut chain)
-            .map_err(Into::into)
-            .and_then(|num_bytes_written| {
-                if num_bytes_written != expected_num_bytes_written {
-                    Err(
-                        err_inc_write(u64::from(num_bytes_written), u64::from(piece_bytes_len))
-                            .into(),
-                    )
-                } else {
-                    Ok(s.sector_id)
-                }
-            })
-            .map(|sector_id| {
-                s.pieces.push(metadata::PieceMetadata {
-                    piece_key,
-                    num_bytes: piece_bytes_len,
-                    comm_p: None,
-                    piece_inclusion_proof: None,
-                });
-
-                sector_id
-            })
-    } else {
-        Err(err_unrecov("unable to retrieve sector from state-map").into())
+        .or_else(|_| provision_new_staged_sector(mgr, &mut sector_builder_state))?;
+
+    let ssm = sector_builder_state
+        .staged
+        .sectors
+        .get_mut(&dest_sector_id)
+        .ok_or_else(|| format_err!("unable to retrieve sector from state-map"))?;
+
+    // TODO: Buffering the piece completely into memory is awful, but each of
+    // the two function calls (add_piece and generate_piece_commitment) accept a
+    // Read. Given that the piece byte-stream is represented by a Read, we can't
+    // read all of its bytes in one function call and then do the same with the
+    // other w/out putting the bytes into an intermediate buffer. A tee reader
+    // would be appropriate here.
+    let mut backing_buffer = vec![];
+    let mut cursor = Cursor::new(&mut backing_buffer);
+
+    std::io::copy(&mut piece_file, &mut cursor)
+        .map_err(|err| format_err!("unable to copy piece bytes to buffer: {:?}", err))?;
```

Comment:
(no need to change, just an fyi)

No need to a cursor for this first part, 

```rust
let mut buffer = Vec::new();
piece_file.read_to_end(&mut buffer)?;

// now create a cursor for passing it to add_piece
let cursor = Cursor::new(&mut buffer);
```

Reply:
word. good tip. in the name of moving forward, i'll just replace this Cursor stuff with code that consumes your new API (in the next PR)